src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -89050,5 +89061,72 @@ class Derived<T> : Base<T> where T : class
             Assert.False(dGoo.Parameters[0].Type.IsNullableType());
             Assert.True(dGoo.Parameters[0].TypeWithAnnotations.NullableAnnotation == NullableAnnotation.Annotated);
         }
+
+        [Fact]
+        [WorkItem(29894, "https://github.com/dotnet/roslyn/issues/29894")]
+        public void TypeOf_03()
+        {
+            var source = @"
+#nullable enable
+class K<T> { }
+class C1
+{
+    object o1 = typeof(int);
+    object o2 = typeof(string);
+    object o3 = typeof(int?);
+    object o4 = typeof(string?);    // 1
+    object o5 = typeof(K<int?>);
+    object o6 = typeof(K<string?>);
+    object o7 = typeof(K<int?>?);   // 2
+    object o8 = typeof(K<string?>?);// 3
+}
+#nullable disable
+class C2
+{
+    object o1 = typeof(int);
+    object o2 = typeof(string);
+    object o3 = typeof(int?);
+    object o4 = typeof(string?);    // 4, 5
+    object o5 = typeof(K<int?>);
+    object o6 = typeof(K<string?>); // 6
+    object o7 = typeof(K<int?>?);   // 7, 8
+    object o8 = typeof(K<string?>?);// 9, 10, 11
+}
```

Comment:
nit: would it also be worth capturing behavior for generic type parameters? (`T`, `TClass`, `TStruct`)